cobra/flux_analysis/phenotype_phase_plane.py

```diff
@@ -374,168 +372,201 @@ def production_envelope(model, reactions, objective=None, c_source=None,
             raise ValueError('cannot calculate yields for objectives with '
                              'multiple reactions')
-        carbon_io = c_input, objective_reactions[0]
+        c_output = objective_reactions[0]
         min_max = fva(model, reactions, fraction_of_optimum=0)
-        grid = [linspace(min_max.minimum[rxn.id], min_max.maximum[rxn.id],
-                         points, endpoint=True) for rxn in reactions]
-        grid_list = list(product(*grid))
-        result = envelope_for_points(model, reactions, grid_list, carbon_io)
-
-    return pd.DataFrame(result)
-
+        min_max[min_max.abs() <
+                model.solver.configuration.tolerances.feasibility] = 0.0
+        points = list(product(*[
+            linspace(min_max.at[rxn.id, "minimum"],
+                     min_max.at[rxn.id, "maximum"],
+                     points, endpoint=True) for rxn in reactions]))
+        tmp = pd.DataFrame(points, columns=[rxn.id for rxn in reactions])
+        grid = pd.concat([grid, tmp], axis=1, copy=False)
+        add_envelope(model, reactions, grid, c_input, c_output)
+    return grid
+
+
+def add_envelope(model, reactions, grid, c_input, c_output):
+    if c_input is not None:
+        input_components = [reaction_components(rxn) for rxn in c_input]
+        output_components = reaction_components(c_output)
+        try:
+            input_weights = [reaction_weight(rxn) for rxn in c_input]
+            output_weight = reaction_weight(c_output)
+        except ValueError:
+            input_weights = []
+            output_weight = []
+    else:
+        input_components = []
+        output_components = []
+        input_weights = []
+        output_weight = []
 
-def envelope_for_points(model, reactions, grid, carbon_io):
-    results = defaultdict(list)
     for direction in ('minimum', 'maximum'):
         sense = "min" if direction == "minimum" else "max"
-        for point in grid:
-            with model:
-                model.solver.objective.direction = sense
-                for reaction, coordinate in zip(reactions, point):
-                    reaction.bounds = coordinate, coordinate
-                model.slim_optimize()
-                if model.solver.status == OPTIMAL:
-                    for reaction, coordinate in zip(reactions, point):
-                        results[reaction.id].append(coordinate)
-                    results['direction'].append(direction)
-                    results['flux'].append(model.solver.objective.value)
-                    if carbon_io[0] is not None:
-                        results['carbon_yield'].append(carbon_yield(carbon_io))
-                        results['mass_yield'].append(mass_yield(carbon_io))
-    for key, value in results.items():
-        results[key] = array(value)
-    if carbon_io[0] is not None:
-        results['carbon_source'] = carbon_io[0].id
-    return results
+        with model:
+            model.solver.objective.direction = sense
```

Comment:
This doesn't use context manager, perhaps better formulate as `model.objective = ..` ? Otherwise I guess will exit with objective changed to max if starting with min..